bokehjs/src/coffee/document.coffee

```diff
@@ -614,6 +614,10 @@ class Document
         new_references[id] = value
 
     Document._initialize_references_json(references_json, old_references, new_references)
+    for id, model of new_references
+      model.attach_document(@)
```

Comment:
The invariant has to be maintained that on inserting a model into the document we attach it exactly once, and that on removing the model from the document we detach it exactly once. This ensures among other things that calls to attach_document and detach_document alternate; so attach_document can unconditionally set up any document-dependent stuff assuming there was no document before, and detach_document can unconditionally tear down any document-dependent stuff.

Things are a little broken right now on the JS side because we don't handle cycles correctly. I believe there's a commented-out test that would be fixed if we switched to doing things the way we do on the Python side. Roughly speaking, the JS side approach is equivalent to reference counting, and the python-side approach is equivalent to a mark-and-sweep GC.

Relevant historical stuff:
 - https://github.com/bokeh/bokeh/pull/3484
 - https://github.com/bokeh/bokeh/commit/9fd682714adf7f7c60f9ab52a24ce35b95ec1951

It looks to me like add_layoutable probably needs similar guarantees as attach/detach. I don't see anything in detach that un-does the add_layoutable in attach - is that an issue? In general I'd be suspicious anytime attach/detach are not "symmetrical", if it's done in one, it needs to be undone in the other.